posts/inside-rust/2024-12-13-this-development-cycle-in-cargo-1.84.md

```diff
@@ -0,0 +1,400 @@
+Cargo includes `RUSTFLAGS` in the fingerprint hash but no in the `-Cextra-filename` hash,
+causing a full rebuild when they change.
+This can be especially problematic when `RUSTFLAGS` differs between the user and their editor running `cargo`.
+For example, some users report they set `--cfg test` in their editor so all `#[cfg(test)]`s are enabled in rust-analyzer.
+
+A previous attempt was made to segregate the cache for `RUSTFLAGS` in
+[#6503](https://github.com/rust-lang/cargo/pull/6503).
+However, Cargo uses the same hash for `-Cextra-filename` and `-Cmetadata`
+so by segregating the cache, the symbol names also become unique.
+In theory, this is a good thing as `RUSTFLAGS` can affect the ABI.
+However, not all `RUSTFLAGS` affect the ABI.
+
+Take [`--remap-path-prefix`](https://doc.rust-lang.org/rustc/command-line-arguments.html#--remap-path-prefix-remap-source-names-in-output) which is suppose to make builds of binaries more reproducible by stripping information specific to a specific build.
+By including this in `-Cmetadata`,
+the binary changes ([#6914](https://github.com/rust-lang/cargo/issues/6914)).
+A special case for this was added [#6966](https://github.com/rust-lang/cargo/pull/6966).
+
+Another case we ran into was [PGO](https://doc.rust-lang.org/rustc/profile-guided-optimization.html).
+With PGO, you create a build with
+[`-Cprofile-generate`](https://doc.rust-lang.org/rustc/codegen-options/index.html#profile-generate)
+and then run it against a benchmark.
+You then feed this back into the build with
+[`-Cprofile-use`](https://doc.rust-lang.org/rustc/codegen-options/index.html#profile-use)
+to improve the optimizations the compiler performs.
+At this point, we reverted 
+[#6503](https://github.com/rust-lang/cargo/pull/6503)
+in
+[#7417](https://github.com/rust-lang/cargo/pull/7417).
+
+In [#8716](https://github.com/rust-lang/cargo/issues/8716),
+[ehuss](https://github.com/ehuss) proposed Cargo track `-Cextra-filename` and `-Cmetadata` separately and only include `RUSTFLAGS` in `-Cextra-filename`.
+
+After some refactoring
+([#14826](https://github.com/rust-lang/cargo/pull/14826))
+and test improvements
+(
+[#14848](https://github.com/rust-lang/cargo/pull/14848),
+[#14846](https://github.com/rust-lang/cargo/pull/14846),
+[#14859](https://github.com/rust-lang/cargo/pull/14859)
+)
+by [epage](https://github.com/epage) and [weihanglo](https://github.com/weihanglo/),
+epage posted [#14830](https://github.com/rust-lang/cargo/pull/14830).
+However, weihanglo found there are still problems with `--remap-path-prefix`: even when using `profile.dev.split-debuginfo="packed"`, the binaries are different because the binary includes `DW_AT_GNU_dwo_name` which points to the debug file which exists per-rlib with `-Cextra-filename` included.
+
+Merging of [#14830](https://github.com/rust-lang/cargo/pull/14830) is blocked until the problem with `--remap-path-prefix` is resolved.
```

Comment:
I believe this was merged?

IIRC, we decided to add a hack for `--remap-path-prefix` specifically?

Reply:
We discussed it on the 3rd and merged on the 6th while this post only covers up to 2024-11-28.

Its less confusing when I get this posted right next to the release.  I'd like to get back to starting to right it before the release but this time there were the holidays.